lib/Scheduling/Problems.cpp

```diff
@@ -157,6 +157,131 @@ LogicalResult CyclicProblem::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// ChainingProblem
+//===----------------------------------------------------------------------===//
+
+LogicalResult ChainingProblem::checkCycleTime() {
+  if (!getCycleTime() || *getCycleTime() <= 0.0f)
+    return getContainingOp()->emitError("Invalid cycle time");
+  return success();
+}
+
+LogicalResult ChainingProblem::checkDelays(OperatorType opr) {
+  auto incDelay = getIncomingDelay(opr);
+  auto outDelay = getOutgoingDelay(opr);
+
+  if (!incDelay || !outDelay)
+    return getContainingOp()->emitError()
+           << "Missing physical delays for operator type '" << opr << "'";
+
+  float zInc = *incDelay;
+  float zOut = *outDelay;
+  float ct = *getCycleTime();
+
+  if (zInc < 0.0f || zOut < 0.0f)
+    return getContainingOp()->emitError()
+           << "Negative physical delays for operator type '" << opr << "'";
+
+  if (zInc > ct)
+    return getContainingOp()->emitError()
+           << "Incoming delay (" << zInc << ") for operator type '" << opr
+           << "' exceeds cycle time (" << ct << ")";
+
+  if (zOut > ct)
+    return getContainingOp()->emitError()
+           << "Outgoing delay (" << zOut << ") for operator type '" << opr
+           << "' exceeds cycle time (" << ct << ")";
```

Comment:
Should these be fatal errors?  It might be useful for these to not be fatal errors, but instead to 'do the best you can'

Reply:
Yes, I think these particular ones (in `checkDelays`) should be errors, because there's no value in trying to fit 10ns operators in 5ns cycles, for example. I'd consider that a bug in the client code constructing the problem.

Also, I'd prefer to keep letting the tests in `verifyPrecedenceInCycle` produce errors, because these hint at bugs in the scheduler.

I think the interesting design decision is in `verifyStartTimeInCycle`. Yes, the cycle time constraint is currently strict (hence the error), but we could introduce a soft constraint to support a 'do the best you can' mode like this:

```c++
prob.setDesiredCycleTime(5.0f);
prob.setMaximumCycleTime(6.0f);
```

We could support instances where at least one of these constraints is set. WDYT?

Reply:
I think in this case, all setMaximumCycleTime() is doing is controlling the policy on emitting an error.  I would decouple the issuing of an error from the scheduling problem.  The critical thing here is that in many systems, the delays have some inaccuracy, so it is really up to the user of the infrastructure whether a failure to schedule is a fatal error.  I would suggest having the infrastructure focus on collecting constraints that aren't satisfied and returning them to the caller and letting the caller figure out whether to issue an error.

I agree that this may be different if you're trying to catch errors in the scheduler.  In those cases, you should indeed emit an error.  There's a bit of a fine line here, I think.

Reply:
Hmm, makes sense. I wonder if I should remove the cycle time altogether from the `ChainingProblem`, and treat it solely as an input to the concrete scheduler (I plan to handle different objective functions in the same way). The problem's verifier would only guarantee the presence of start-times-in-cycle, and that these are consistent w.r.t. the problem's dependences.

This gives the client maximum flexibility in inspecting & accepting the solution as needed. The client controls what it gets (strict adherence to cycle time, or best-effort) through its choice of algorithm.

Reply:
So before this PR goes totally stale 😉: I think the `ChainingProblem` still makes sense without the cycle time constraint, and I just pushed the corresponding changes.

@mikeurbach, @stephenneuendorffer: would you be ok with this for now? I would then look into implementing some scheduler support for this problem, and test everything out in @rachitnigam use-case in Calyx.

Reply:
This looks like a good place to start off work for the Calyx use-case which would involve breaking up large `calyx.comb_group` blocks into separate `calyx.group` which perform combinational computations and store them into registers. It would be interesting to see how we can incorporate `Problem` instances directly into the transformation over Calyx programs and generate new groups.